### PR TITLE
Add support for offset load/store

### PR DESCRIPTION
**Pull request by YixingZhang007**
Add a new form of load/store operations for cooperative matrices that accepts two separate arguments: the row index and the column index. Unlike the original approach requiring a pointer to the matrix base, this new form of load/store operations is expected to yield better optimized code on 2dblock read/write instructions on PVC.

CapabilityCooperativeMatrixOffsetInstructionsINTEL = 6238
OpCooperativeMatrixLoadOffsetINTEL = 6239
OpCooperativeMatrixStoreOffsetINTEL = 6240

### CODE REVIEW

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_bfloat16_offset.ll
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_bfloat16_offset.ll:284]: Please shorten the tests - we won't need all of these metadata

Comment (YixingZhang007) [reply test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_bfloat16_offset.ll]: Thanks for the suggestions! I have updated the test.

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3713]:
```suggestion
class SPIRVCooperativeMatrixOffsetInstructionsINTELInstBase
```

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3727]: same

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll:1]: Lets use KHR/SPV_KHR_cooperative_matrix test as the template...

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll:28]: ...so here we would have CooperativeMatrixMulAddKHR instruction.

Reason: JointMatrixMadINTEL will be deprecated. And real-life SYCL joint_matrix workflows with switch to cooperative matrix KHR when https://github.com/intel/llvm/pull/16045 is merged

**Review by MrSidims — COMMENTED** [review]
2 nits, other than that the patch is LGTM

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h
Comment (YixingZhang007) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h]: Thanks! I have fixed this :)

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h
Comment (YixingZhang007) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h]: I have also fixed this.

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll
Comment (YixingZhang007) [reply test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll]: Thanks for the suggestions! I have updated the test :)

**Review by MrSidims — APPROVED** [review]

**Comment by YixingZhang007** [comment]
@vmaksimo I just did some code clean up. Could you please help review this PR? Thanks for helping :)

**Comment by MrSidims** [comment]
Spec is available as rev 17 of https://github.com/intel/llvm/pull/12497
@vmaksimo may I ask you to take a look at CI issues?

**Comment by vmaksimo** [comment]
CI issue was fixed in main by 4a85dd8, please rebase on top to get CI passing

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h
Comment (vmaksimo) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3750]: Please verify that the last optional operand `Memory Operand` is really translated as `Literal` to SPIR-V (there are no checks for that in test, so the bug can be hidden). If no, please apply my suggestion, it should work this way if I'm not misunderstanding 
```suggestion
_SPIRV_OP(CooperativeMatrixLoadOffset, true, 8, true, 5)
_SPIRV_OP(CooperativeMatrixStoreOffset, false, 7, true, 6)
```

Comment (YixingZhang007) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h]: Hello Viktoria! Thanks so much for the suggestion! I have checked that the last operand `Memory Operand` is not translated to `Literal`. I have applied the suggested change :)

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll
Comment (vmaksimo) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll:31]: Small naming suggestion to change `Ptr2/Ptr3` to something like "load result" to avoid confusion with the pointer operand.
Also seems that `CooperativeMatrixMulAddKHR` last operand should also be literal (will it be simply `64` or a Cooperative Matrix Operands	enum value?).
```suggestion
; CHECK-SPIRV: CooperativeMatrixLoadOffsetINTEL [[#MatTy2]] [[#Load2:]] [[#]] [[#Index2:]] [[#]] [[#Const0]] [[#Const128]] 0
; CHECK-SPIRV: CooperativeMatrixLoadOffsetINTEL [[#MatTy3]] [[#Load3:]] [[#]] [[#]] [[#]] [[#Const2:]] [[#Const256:]] 0
; CHECK-SPIRV: CooperativeMatrixMulAddKHR [[#MatTy1]] [[#]] [[#Load2]] [[#Load3]] [[#Result:]] 64
```

**Review by vmaksimo — APPROVED** [review]
LGTM. 
Applying the suggestion is not required to get this merged.

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll
Comment (YixingZhang007) [reply test/extensions/INTEL/SPV_INTEL_joint_matrix/joint_matrix_load_store_offset.ll]: Hello Viktoria! Thanks for the suggestion! I have applied the changes :)

**Comment by YixingZhang007** [comment]
@vmaksimo Could you please help merge this PR or if I need to get approval from anyone else? Thanks!

**Pull request merged**
